lib/test.js

```diff
@@ -93,7 +93,21 @@ Test.prototype.run = function () {
         this.timeoutAfter(this._timeout);
     }
     this.emit('prerun');
-    this._cb(this);
+    var callbackReturn = this._cb(this);
+
+    if (
+        typeof Promise === 'function' &&
+        callbackReturn &&
+        typeof callbackReturn.then === 'function'
+    ) {
+        var self = this;
+        Promise.resolve(callbackReturn).catch(function onError(err) {
```

Comment:
Note: another alternative implementation is:
```jsx
new Promise(function (resolve) {
  resolve(self._cb(self));
}).catch(…);
```
which would ensure that *sync* errors don't kill the whole suite but instead fail the test (which mimics other test runners), and would make sync and async errors behave the same.

Reply:
This creates zalgo behavior where only exceptions thrown synchronously are caught and exceptions in callbacks go to uncaught handler.

With the `async/await` implementation, if the application developer uses promises and `await` all exceptions, whether synchronous or after `await` are caught in the rejected promise. This only holds true if they don't use any callbacks and rely on promises / `events.once` / `util.promisify`

Reply:
It's always the case - whether using async/await or not - that if you use async callbacks and they throw an exception, it won't be caught. That's not creating zalgo behavior, it's just unavoidable.

I don't think it's zalgo that sync exceptions fail the test while non-promise async ones continue to have the same behavior.